keps/sig-api-machinery/2855-automated-stored-versions/README.md

```diff
@@ -0,0 +1,580 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The [kube-storage-version-migrator] supports migrating existing resources to a new storage version but it does not modify a CRD's `status.storedVersions` field after a successful migration, instead requiring that the cluster administrator update this field manually. As adoption for CRDs continues to increase and new versions of those CRDs are released this design decision will place a non-trivial amount of work on cluster administrators.
+
+Historically, the [kube-storage-version-migrator] has not managed modifying the `status.storedVersion` field because the CRD's storedVersion could change mid-migration unbenownst to the migrator due to slow/failed watches. This KEP proposes a solution that would allow the [kube-storage-version-migrator] to safely modify a CRD's `status.storedVersions` after a successful migration, reducing the burden on cluster administrators.
+
+[kube-storage-version-migrator]: https://github.com/kubernetes-sigs/kube-storage-version-migrator
+
+## Motivation
+
+It is desirable for the [kube-storage-version-migrator] to modify a CRD's `status.storedVersions` array after a successful migration rather than relying on a cluster admin to manually perform the same steps.
+
+### Goals
+
+- Allow the [kube-storage-version-migrator] to modify a CRD's `status.storedVersions` after a successful migration.
+
+## Proposal
+
+In order for the [kube-storage-version-migrator] to safely modify a CRD's `status.storedVersions` field after a successful migration, this KEP proposes that the [kube-storage-version-migrator] introduce a [Validating Admission Webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/) that prevents changes to a CRD's `spec.versions[*].storage` fields during a migration.
+
+The validating admission webhook would need to:
```

Comment:
Several details to be filled in at least in the implementation section:

Will it be mandatory to have the webhook? Current deployments won't have it. And I guess you want the deployment of the admission webhook as a separate manifest (as it heavily depends on the avaiilable CA provisioner in the cluster). Should the ksvm process stop launching if the admission webhook is not installed or not provisioned with a CA (yet) ?

Which error behaviour will this use? Fail closed?

Reply:
ic, error behaviour is defined in https://github.com/kubernetes/enhancements/pull/2856/files#diff-b0cf963682236f4ebe7694cc958b7215f37ca03b8f37b224fba35f8fd483ac7cR178.